libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -1006,21 +1008,27 @@ public XContentBuilder copyCurrentStructure(XContentParser parser) throws IOExce
 
     /**
      * Sets a version used for serialising a response compatible with a previous version.
+     * @param compatibleVersion - indicates requested a version of API that the builder will be creating
      */
-    public XContentBuilder withCompatibleMajorVersion(byte compatibleMajorVersion) {
-        assert this.compatibleMajorVersion == 0 : "Compatible version has already been set";
-        if (compatibleMajorVersion == 0) {
-            throw new IllegalArgumentException("Compatible major version must not be equal to 0");
+    public XContentBuilder withCompatibleVersion(CompatibleVersion compatibleVersion) {
+        assert this.restCompatibilityVersion == null : "Compatible version has already been set";
+        if (compatibleVersion == null) {
+            throw new IllegalArgumentException("Compatible major version must not be null");
         }
-        this.compatibleMajorVersion = compatibleMajorVersion;
+        this.restCompatibilityVersion = compatibleVersion;
         return this;
     }
 
     /**
-     * Returns a version used for serialising a response compatible with a previous version.
+     * Returns a version used for serialising a response.
+     * @return a compatible version
      */
-    public byte getCompatibleMajorVersion() {
-        return compatibleMajorVersion;
+    public CompatibleVersion getRestCompatibilityVersion() {
+        return restCompatibilityVersion;
+    }
+
+    public boolean useCompatibility(CompatibleVersion codeChangeVersion) {
```

Comment:
I agree, I personally prefer to replace boolean expressions with named methods, but in this case it would be an overdo.
I also realised that we can apply the same 'getter' style to parser